File: python/0589.N叉树的前序遍历.py
class Solution:
    # 1. O(n) t:68ms(7%) O(n) m:16.9M(44%) 迭代
    def preorder(self, root: 'Node') -> List[int]:
        stack, ans = [root], []
        while stack:
            node = stack.pop(0)
            if node:
                ans.append(node.val)
                stack = node.children + stack
            else:
                continue
        return ans
    
    # [题解]2. O(n) t:68ms(7%) O(1) m:17.3M(5%) 递归
    def preorder(self, root: 'Node') -> List[int]:
        return ([root.val] + [x for node in root.children for x in self.preorder(node)]) if root else []

File: cpp/0589.N叉树的前序遍历.cpp
class Solution {
public:
    // [题解]1. O(n) t:12ms(95%) O(1) m:11.4M(22%) 递归
    vector<int> ans;
    vector<int> preorder(Node* root) {
        dfs(root, ans);
        return ans;
    }

    void dfs(const Node* root, vector<int> &ans) {
        if (root == nullptr) return ;
        ans.emplace_back(root->val);
        for (auto &node : root->children){
            dfs(node, ans);
        }
    }

    // [题解]2. O(n) t:16ms(75%) O(n) m:11M(93%) 迭代
    vector<int> preorder(Node* root) {
        vector<int> ans;
        if (root == nullptr) return ans;
        stack<Node *> st;
        st.emplace(root);
        while (!st.empty()) {
            Node *node = st.top();
            st.pop();
            ans.emplace_back(node->val);
            for (auto it = node->children.rbegin(); it != node->children.rend(); it++) {
                st.emplace(*it);
            }
        }
        return ans;
    }
};